### PR TITLE
Building A React App With TypeScript

### LINKED ISSUE #3262: How To Build A React App WithTypeScript
### Introduction paragraph (2-3 paragraphs):
React is one of the most popular and best front-end libraries as of today this has made a lot of developers make it their primary library used for creating front-end applications. Typescript on the other hand is a strict and statically typed programming language which makes our javascript code more predictable. As developers, our goal is not just to create products but to create products that we can always predict their behavior, this is one of the main reasons why Typescript was created. With Typescript, we can create predictable react applications. 

This article will demonstrate how to build a simple application while explaining the core concepts of React and Typescript.

### Key takeaways: 
1. How to create a react app with typescript
2. Declaring types in typescript
3.  State and props in functional components
4. State and Props in class based components
5. Event types

### References: 
1. [typescript - react](https://www.typescriptlang.org/docs/handbook/react.html)
2. [react docs](https://reactjs.org/)
3. [typescript docs](https://www.typescriptlang.org/)

### PR DESCRIPTION
**Pull request by Babatunde13**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #3262 

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by marienjus** [comment]
Hi I'll be your peer reviewer.

File: content/articles/building-a-react-app-with-typescript/index.md [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+images:
+
+  - url: /engineering-education/building-a-react-app-with-typescript/react-ts.png
+    alt: React App Using TypeScript
+---
+
+## Introduction
+
+![](react-ts.png)
+
```
Comment (marienjus) [inline content/articles/building-a-react-app-with-typescript/index.md:21]: Check Your image paths

File: content/articles/building-a-react-app-with-typescript/index.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+We use Create React App to set up our react application that way we do not need to worry about all the webpack and babel stress, as you might have guessed to set up a react typescript application we'd still use create-react-app just with some extra flags. In our terminal we'd type the following to bootstrap our application
+
+```bash
+npx create-react-app todo-app --template typescript
+```
+
+This command will create a react typescript application for us then we can set up the server as we will normally. Our app structure should look like this
+
+![](./filestructure.png)
+
```
Comment (marienjus) [inline content/articles/building-a-react-app-with-typescript/index.md:45]: Hey you should use `/section-engineering-education/building-a-react-app-with-typescript/imageName` as the path. Kindly change every instance. This is your first article and therefore make sure to go through the [new contributors guide](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/UPLOAD_INSTRUCTIONS.md)

**Comment by marienjus** [comment]
Seems you made the commit directly from your forked master branch?

**Comment by marienjus** [comment]
This article needs a hero image.

File: content/articles/building-a-react-app-with-typescript/index.md (hunk 1 above)
Comment (Babatunde13) [reply content/articles/building-a-react-app-with-typescript/index.md:21]: Thanks for the feedback, I'd work on that now.

File: content/articles/building-a-react-app-with-typescript/index.md (hunk 2 above)
Comment (Babatunde13) [reply content/articles/building-a-react-app-with-typescript/index.md]: 👍

**Comment by marienjus** [comment]
The format and flow is a smooth sail :rocket:
@Babatunde13 I will now be taking you through a tech review.

**Comment by Babatunde13** [comment]
> The format and flow is a smooth sail  Of course, mentioning React as the best frontend library will always be controversial 
> @Babatunde13 I will now be taking you through a tech review.

Well, React is definitely not the best but then we know it has the highest preference among developers, I guess that's what makes it the best at the end of the day :blush:

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by marienjus** [comment]
Your last article has update has minor Grammar errors. Kindly use Grammarly to remove extra spaces so that we can proceed to a final review.

File: content/articles/building-a-react-app-with-typescript/index.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+const data: something = "Hey"
+
+```
+
+As we can see, we were able to define a type for more complex data, and we can use this idea to model much more complex data. From this example, we might be able to see a difference between type and interface.
+
+
+### More Details ABout Our Application
+In our `App.tsx` file, we need to delete everything inside the function and return a simple text
```
Comment (marienjus) [inline content/articles/building-a-react-app-with-typescript/index.md:77]: This too. *About

**Comment by hectorkambow** [comment]
> > The format and flow is a smooth sail  Of course, mentioning React as the best frontend library will always be controversial
> > @Babatunde13 I will now be taking you through a tech review.
> 
> Well, React is definitely not the best but then we know it has the highest preference among developers, I guess that's what makes it the best at the end of the day 😊

Statements like this require a source. - FYI

**Comment by hectorkambow** [comment]
Overall nice work. Made some changes to the format and flow.

Please go over any edits made by the reviewer or myself to improve future reviews. 

@Babatunde13 

cc @marienjus

**Pull request merged**